src/software/geom/algorithms/nearest_neighbor_search.hpp

```diff
@@ -0,0 +1,33 @@
+#pragma once
+
+#include <vector>
+
+/**
+ * Finds the neighbors of a candidate based on a comparator and radius.
+ * Neighbors on the edge of the radius are included.
+ *
+ * @tparam T generic type for the candidate and its neighbors
+ * @tparam Func type for the function comparator. Must be in the form `double f(const T&,
+ * const T&)`
+ * @param candidate cannot be null
+ * @param neighbors a list of T objects neighbors, which may contain the candidate object
+ * itself
+ * @param radius the maximum distance from a candidate object
+ * @param comparator function that measures the distance between a candidate and a
+ * neighbor
+ * @return a list representing a subset of neighbors
+ */
+template <class T, typename Func>
+std::vector<T> distanceFunc(const T& candidate, const std::vector<T>& neighbors,
+                            double radius, Func comparator)
```

Comment:
```suggestion
                            double radius, Func distanceFunc)
```
as per nimas comment

Reply:
Fixed